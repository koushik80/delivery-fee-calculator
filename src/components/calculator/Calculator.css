.calculator {
    width: 90%;
    margin: 2rem auto;
    display: flex;
    align-items: center;
    justify-content: center;
    margin-top: 100px;
}

.calculator-div {
    padding: 4rem;
    border-radius: 20px;
    box-shadow: 0 0 10px #ffffff;
    background: linear-gradient(-55deg, #e63504, #ece3e6, #55afcf, #4cdab9, #d8d889, #9c87ef, #FE676E);
    background-size: 400% 400%;
    animation: gradient 15s ease infinite;
}

@keyframes gradient {
  0% {
    background-position: 0% 50%;
  }
  50% {
    background-position: 100% 50%;
  }
  100% {
    background-position: 0% 50%;
  }
}
.calculator-title {
    font-size: 2rem;
    color: #ffffff;
    margin-bottom: 1rem;
    font-weight: 700;
}
.calculator-form-div {
    display: flex;
    align-items: center;
    gap: 16px;
    margin-top: 2rem;
}

.form-input {
    display: flex;
    flex-direction: column;
    gap: 12px;
    justify-content: space-between;
}

.form-label {
    gap: 28px;
}

.form-input label, span {
    font-size: 1.2rem;
    font-weight: 600;
}

.form-input input {
  font-size: 1rem;
  padding: 4px 12px;
  padding: 10px;
	width: 82%;
	outline: none;
	border: 1px solid #bbb;
  margin-right: 6px;
  border-radius: 2px;
}

.form-input img {
  width: 1rem;
  height: 1rem;
}

.submit-btn {
    display: inline-block;
    font-weight: 700;
    text-align: center;
    text-transform: uppercase;
    white-space: nowrap;
    vertical-align: middle;
    user-select: none;
    border: 1px solid transparent;
    padding: 0 50px;
    font-size: 14px;
    line-height: 50px;
    border-radius: 8px;
    color: #FFFFFF;
    cursor: pointer;
    z-index: 5;
    transition: 1s;
    background-color: #1dadf4;
    width: 100%;
    position: relative;
    z-index: 10;
    overflow: hidden;
    letter-spacing: 1.5px;
    margin-top: 2rem;
    -webkit-border-radius: 8px;
    -moz-border-radius: 8px;
    -ms-border-radius: 8px;
    -o-border-radius: 8px;
}

.submit-btn::before {
    position: absolute;
    content: "";
    background-color: #7aaafd;
    width: 100%;
    height: 0%;
    left: 50%;
    top: 50%;
    border-radius: 33px;
    -webkit-transform: translate(-50%, -50%) rotate(-45deg);
    transform: translate(-50%, -50%) rotate(-45deg);
    z-index: -1;
    -webkit-transition: all 1000ms ease;
    transition: all 1.5s ease;
    opacity: 1;
}

.submit-btn:hover {
    background-color: #FFFFFF;
    /* background-image: linear-gradient(to right, #FFF, #FFF); */
    color: #1dadf4;
    border-color: #1dadf4;
}

.submit-btn:hover::before {
    height: 400px;
    opacity: 0;
}

.delivery-price {
    font-size: 1.2rem;
    margin-top: 1.5rem;
    font-weight: bold;
}

@media screen and (max-width: 546px) {

.form-input {
    display: flex;
    flex-direction: column;
    gap: 12px;
    justify-content: space-between;
}

.form-label {
    gap: 22px;
}

.form-input label, span {
   font-size: .8rem;
   font-weight: 600;
}

.form-input input {
    font-size: .8rem;
    padding: 8px;
	  width: 80%;
	  outline: none;
	  border: 1px solid #bbb;
    margin-right: 6px;
    border-radius: 2px;
}

.calculator-div {
    padding: 1rem;
}

}