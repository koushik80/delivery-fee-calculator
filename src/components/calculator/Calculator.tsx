import { useState } from 'react'
import calendarIcon from './../../images/calendar.png'
import './Calculator.css'
import { handleCartValue, handleDeliveryDistance, handleNumberOfItems, handleTime } from '../../utils/utils'

const Calculator = () => {
  const [cartValue, setCartValue] = useState<number>()
  const [deliveryDistance, setDeliveryDistance] = useState<number>()
  const [amountOfItems, setAmountOfItems] = useState<number>()
  const [time, setTime] = useState<Date>()
  const [deliveryCharge, setDeliveryCharge] = useState<number>(0)

  const handleDeliveryCharge = (e: any) => {
    e.preventDefault()
    let surCharge = handleCartValue(Number(cartValue))
    let deliveryCharge = handleDeliveryDistance(Number(deliveryDistance))
    let itemsSurCharge = handleNumberOfItems(Number(amountOfItems))
    const isRushTime = handleTime(String(time))

    let finalDeliveryCharge = surCharge + deliveryCharge + itemsSurCharge


    if (isRushTime) {
      finalDeliveryCharge = finalDeliveryCharge * 1.2 //during rush time delivery rate is 1.2 times higher than the regular delivery fee
    }

    if (Number(cartValue) === 100 || Number(cartValue) > 100) {
      finalDeliveryCharge = 0  // when the cart value is 100€ or more than that no charge will be added
    }

    if(finalDeliveryCharge > 15) {
      finalDeliveryCharge = 15 //delivery charge won't cross more than 15€
    }

    setDeliveryCharge(finalDeliveryCharge)
  }

  return (
    <div className="calculator">
      <div className="calculator-div">
        <h4 className="calculator-title">Calculator</h4>
        <form className="calculator-form">
          <div className="calculator-form-div">
            <div className="form-input form-label">
              <div>
                <label htmlFor="cartValue">Cart Value</label>
              </div>
              <div>
                <label htmlFor="deliveryDistance">Delivery distance</label>
              </div>
              <div>
                <label htmlFor="amountOfItems">Amount of items</label>
              </div>
              <div>
                <label htmlFor="time">Time</label>
              </div>
            </div>
            <div className="form-input">
              <div>
                <input
                  required
                  type="number"
                  id="cartValue"
                  min="0"
                  onChange={({ target }) => setCartValue(Number(target.value))}
                  onBlur={({ target }) => setCartValue(Number(target.value))}
                />
                <span>€</span>
              </div>
              <div>
                <input
                  required
                  type="number"
                  id="deliveryDistance"
                  onChange={({ target }) =>
                    setDeliveryDistance(Number(target.value))
                  }
                  onBlur={({ target }) =>
                    setDeliveryDistance(Number(target.value))
                  }
                />
                <span>m</span>
              </div>
              <div>
                <input
                  required
                  type="number"
                  min="0"
                  id="amountOfItems"
                  onChange={({ target }) =>
                    setAmountOfItems(Number(target.value))
                  }
                  onBlur={({ target }) =>
                    setAmountOfItems(Number(target.value))
                  }
                />
              </div>
              <div>
                <input
                  type="datetime-local"
                  id="time"
                  onChange={({ target }) => setTime(new Date(target.value))}
                  onBlur={({ target }) => setTime(new Date(target.value))}
                  required
                />
                <img src={calendarIcon} alt="Calendar" />
              </div>
            </div>
          </div>
          <button
            onClick={handleDeliveryCharge}
            type="submit"
            className="submit-btn"
          >
            Calculate delivery price
          </button>
        </form>

        <div className="delivery-price">
          <p>Delivery price: {deliveryCharge} € </p>
        </div>
      </div>
    </div>
  )
}

export default Calculator