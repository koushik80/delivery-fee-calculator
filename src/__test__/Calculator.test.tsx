import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import Calculator from '../components/calculator/Calculator'
import {
  handleCartValue,
  handleDeliveryDistance,
  handleNumberOfItems,
  handleTime,
} from './../utils/utils'

describe('Calculator component test', () => {
  let wrapper: any

  beforeEach(() => {
    wrapper = shallow(<Calculator />)
  })

  it('Render correctly initial document', () => {
    const tree = renderer.create(<Calculator />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render the calculator form', () => {
    expect(wrapper.find(".calculator-form")).toHaveLength(1)
  })

  it('should render the 4 input fields', () => {
    expect(wrapper.find("input")).toHaveLength(4)
  })

  it('should render the submit button', () => {
    expect(wrapper.find(".submit-btn")).toHaveLength(1)
  })

  it('should render the delivery price element', () => {
    expect(wrapper.find(".delivery-price")).toHaveLength(1);
  })

  describe('handleCartValue', () => {
    test("should return 1.1 if value is less than 10", () => {
      expect(handleCartValue(5)).toEqual(1.1)
    })

    test('should return 0 if value is greater than 10', () => {
      expect(handleCartValue(11)).toEqual(0)
    })
  })

  describe('handleDeliveryDistance', () => {
    test("should return 1 if value is less than 500", () => {
      expect(handleDeliveryDistance(499)).toEqual(1);
    })

    test('should return 2 if value is between 500 and 1000', () => {
      expect(handleDeliveryDistance(750)).toEqual(2);
    })
    test('should return 3 if value is greater than 1000', () => {
      expect(handleDeliveryDistance(1001)).toEqual(3);
    })
  })

  describe('handleNumberOfItems', () => {
    it('should return 0 if items is less than or equal to 4', () => {
      expect(handleNumberOfItems(4)).toEqual(0)
    })

    it('should return 1 if items is greater than 4', () => {
      expect(handleNumberOfItems(5)).toEqual(0.5)
    })
  })

  describe('handleTime', () => {
    it('should return false if day is not Friday or time is not between 15 and 19', () => {
      expect(handleTime('Sat Aug 29 2020 14:00:00 GMT-0500')).toBe(false);
    })
  })
})