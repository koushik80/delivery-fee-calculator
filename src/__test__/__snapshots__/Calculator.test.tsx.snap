// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component test Render correctly initial document 1`] = `
<div
  className="calculator"
>
  <div
    className="calculator-div"
  >
    <h4
      className="calculator-title"
    >
      Calculator
    </h4>
    <form
      className="calculator-form"
    >
      <div
        className="calculator-form-div"
      >
        <div
          className="form-input form-label"
        >
          <div>
            <label
              htmlFor="cartValue"
            >
              Cart Value
            </label>
          </div>
          <div>
            <label
              htmlFor="deliveryDistance"
            >
              Delivery distance
            </label>
          </div>
          <div>
            <label
              htmlFor="amountOfItems"
            >
              Amount of items
            </label>
          </div>
          <div>
            <label
              htmlFor="time"
            >
              Time
            </label>
          </div>
        </div>
        <div
          className="form-input"
        >
          <div>
            <input
              id="cartValue"
              min="0"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="0"
              required={true}
              type="number"
            />
            <span>
              €
            </span>
          </div>
          <div>
            <input
              id="deliveryDistance"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="0"
              required={true}
              type="number"
            />
            <span>
              m
            </span>
          </div>
          <div>
            <input
              id="amountOfItems"
              min="0"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="0"
              required={true}
              type="number"
            />
          </div>
          <div>
            <input
              id="time"
              onBlur={[Function]}
              onChange={[Function]}
              required={true}
              type="datetime-local"
            />
            <img
              alt="Calendar"
              src="calendar.png"
            />
          </div>
        </div>
      </div>
      <button
        className="submit-btn"
        onClick={[Function]}
        type="submit"
      >
        Calculate delivery price
      </button>
    </form>
    <div
      className="delivery-price"
    >
      <p>
        Delivery price: 
        0
         € 
      </p>
    </div>
  </div>
</div>
`;
